[MODEL] New model support for naver-hyperclovax/HyperCLOVAX-SEED-Vision-Instruct-3B

bigshanedogg opened the pull request:
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.


## Purpose
- Add new model for [naver-hyperclovax/HyperCLOVAX-SEED-Vision-Instruct-3B](https://huggingface.co/naver-hyperclovax/HyperCLOVAX-SEED-Vision-Instruct-3B)
  - FIX #19963 
  - FIX #21275

## Test Plan

### offline inference
```bash
PYTHONPWD=${PWD} python examples/offline_inference/vision_language.py \
        --model-type hyperclovax_seed_vision \
        --modality image \
        --num-prompts 4
```

### online serving
```bash
python -m vllm.entrypoints.openai.api_server \
    --model naver-hyperclovax/HyperCLOVAX-SEED-Vision-Instruct-3B \
    --trust_remote_code \
    --chat-template-content-format "openai" \
    --max-model-len 4096 \
    --max-num-seqs 8
```

## Test Result

### offline inference
```
--------------------------------------------------
A tower is in the background behind a tree with pink blossoms.
--------------------------------------------------
A tower is seen behind a tree with pink blossoms.
--------------------------------------------------
A tower is seen behind a tree with pink blossoms.
--------------------------------------------------
A tower is seen behind a tree with pink blossoms.
--------------------------------------------------
```

### online serving
```
# client
messages = [
    {
        "role": "user",
        "content": [
            {
                "type": "image_url",
                "image_url": {"url": "https://upload.wikimedia.org/wikipedia/commons/d/da/2015_Kaczka_krzy%C5%BCowka_w_wodzie_%28samiec%29.jpg"},
            },
            {
                "type": "text", 
                "text": "Describe about the image.",
            },
        ],
    },
]

payload = {
    "messages": messages,
    "max_tokens": 256,
    "repetition_penalty": 1.0,
}

response = requests.post(
    f'{url}/v1/chat/completions', 
    headers=headers,
    json=payload,
)
response = response.json()
response["choices"][-1]["message"]["content"]
```

```
# result
This is a duck with its head up, floating on a water surface.  The duck is green and yellow in color.  The duck is located in a body of water.  The duck's reflection on the water's surface can be seen below it.  The floating duck overlaps a large area with light blue water.  The reflection of the duck being darker against the bright blue background water is clearly visible below the duck.  The duck's body takes up the space from the left to a little more than one-half way to the right.  The duck's back can be seen go all the way up toward the top of the image, and the duck's head is raised up to just above the top of the image.  The body of the duck slopes down from top to bottom of the image.  The duck is located towards the top left side of the image.  The reflection of the duck below the actual duck is a little shorter than the duck itself.
```

## (Optional) Documentation Update
- [x] updated `supported_models.md`
- [x] updated `examples/offline_inference/vision_language.py`
- [x] updated `examples/offline_inference/vision_language_multi_image.py`
- [x] updated `tests/models/multimodal/processing/test_common.py`

<!--- pyml disable-next-line no-emphasis-as-heading -->

DarkLight1337 commented on requirements/test.txt:
```diff
@@ -305,7 +305,7 @@ mbstrdecoder==1.1.3
 mdurl==0.1.2
     # via markdown-it-py
-mistral-common==1.6.2
+mistral-common==1.7.0
```
Can you merge from main? This file shouldn't be changed

Isotr0py commented on examples/offline_inference/hyperclovax_seed/hyperclovax_seed_vision_3b_instruct.py:
We can move the content of this file to `examples/offline_inference/vision_language.py` and `examples/offline_inference/vision_language_multi_image.py`.

Isotr0py commented on vllm/model_executor/models/hyperclovax_vision.py:
```diff
@@ -0,0 +1,1220 @@
+        self.text_config = text_config
+
+        # etc
+        self.use_liger = kwargs.pop("use_liger", False)
+        self.use_fused_ce = kwargs.pop("use_fused_ce", False)
+        self.use_meansum_loss = kwargs.pop("use_meansum_loss", False)
+        self.freeze_before_sampler = kwargs.pop("freeze_before_sampler", False)
+        self.use_turnmeansum_loss = kwargs.pop("use_turnmeansum_loss", False)
+        self.vision_input_chunk_size = kwargs.pop("vision_input_chunk_size",
+                                                  None)
+        self.is_safetensor_save = kwargs.get("is_safetensor_save", True)
```
```suggestion
```
I think these fields are unused.

Isotr0py commented on vllm/model_executor/models/hyperclovax_vision.py:
```diff
@@ -0,0 +1,1220 @@
+                                               dim=0)
+
+        visual_token_idx = 0 if "siglip" in self.vision_config.model_type else 1
+        context_vision_model = torch.no_grad(
+        ) if self.vision_model_use_no_grad else contextlib.nullcontext()
+        with context_vision_model:
```
I think there is no need to add redundant `torch.no_grad` context here.

Isotr0py commented on vllm/model_executor/models/hyperclovax_vision.py:
```diff
@@ -0,0 +1,1220 @@
+        return reduction
+
+    def _init_vision_config(
+        self,
+        config,
+    ):
+        vision_model_type = config.vision_config.model_type
+        if vision_model_type in CONFIG_MAPPING:
+            vision_config = CONFIG_MAPPING[vision_model_type](
+                **config.vision_config.to_dict())
+            vision_config.auto_map = {}
+        else:
+            if config.vision_model_name_or_path is not None:
+                vision_config = AutoConfig.from_pretrained(
+                    config.vision_model_name_or_path, trust_remote_code=True)
+            elif config.vision_config._name_or_path is not None:
+                vision_config = AutoConfig.from_pretrained(
+                    config.vision_config._name_or_path, trust_remote_code=True)
+            else:
+                raise ValueError("vision_config is not defined")
+
+        vision_config.anyres = config.anyres
+        vision_config.max_num_grids = config.max_num_grids
+        return vision_config
+
+    def _init_text_config(
+        self,
+        config,
+    ):
+        if hasattr(config, "text_config") and config.text_config is not None:
+            model_type = config.text_config.model_type
+            text_config = CONFIG_MAPPING[model_type](
+                **config.text_config.to_dict())
+        else:
+            raise ValueError("text_config is not defined")
+        if config.text_config.model_type in ["llama", "hyperclovax", "gpt2"]:
+            text_config._attn_implementation = "sdpa"
+        if text_config.model_type != "hyperclovax":
+            text_config.logits_scaling = 1.0
+        return text_config
```
Perhaps move the custom config related methods to `vllm/transformers_utils/configs`?

bigshanedogg commented:
I've addressed the feedback you left in the comments and updated the PR description accordingly.

- Removed unnecessary lines
- Moved the examples to a more appropriate location
- Rebased onto the latest main

(cc. @Isotr0py , @DarkLight1337)

DarkLight1337 commented on tests/models/registry.py:
```diff
@@ -196,6 +196,9 @@ def check_available_online(
                                              trust_remote_code=True),
     "HunYuanMoEV1ForCausalLM": _HfExamplesInfo("tencent/Hunyuan-A13B-Instruct",
                                                trust_remote_code=True),
+    "HCXVisionForCausalLM": _HfExamplesInfo(
```
Can you also update `tests/models/multimodal/processing/test_common.py` with this model?

DarkLight1337 commented on vllm/model_executor/models/hyperclovax_vision.py:
```diff
@@ -0,0 +1,1160 @@
+        return self.language_model.compute_logits(hidden_states,
+                                                  sampling_metadata)
+
+    def sample(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> Optional[SamplerOutput]:
+        return self.language_model.sample(logits, sampling_metadata)
```
Can also remove this method

DarkLight1337 commented on tests/models/multimodal/processing/test_common.py:
```diff
@@ -260,59 +261,60 @@ def _test_processing_correctness_one(
 ])
 @pytest.mark.parametrize("hit_rate", [0.3, 0.5, 1.0])
-@pytest.mark.parametrize("num_batches", [32])
+@pytest.mark.parametrize("num_batches", [1])
```
Are you working on fixing the multi-modal processor for this model?

bigshanedogg replied:
I'm currently looking into a `pytest` failure related to the comment you left.  
The following error is occurring:

```
RuntimeError: Expected there to be 3 prompt updates corresponding to 3 video items, but instead found 0 prompt updates!
```

However, everything seems to be working correctly in `examples/offline_inference`, including both `vision_language.py` (for both image and video) and `vision_language_multi_image.py`, as well as in `online_serving`.

DarkLight1337 replied:
Running the examples is not sufficient to ensure the correctness because of the small batch size involved. You should step through the test and see what is happening.

bigshanedogg replied:
Sure, I'll complete the changes and the rebase, then mention you again.

bigshanedogg replied:
I've updated test_common.py.

Additionally, I’ve requested an update on the Hugging Face hub to remove the dependency on transformers==4.52.* for naver-hyperclovax/HyperCLOVAX-SEED-Vision-Instruct-3B.
It seems that the recently added models on the main branch were developed against transformers==4.53.*, so this change is intended to align with that version.

(cc. @DarkLight1337)

dhkim0225 replied:
```
Additionally, I’ve requested an update on the Huggingface hub to remove the dependency on transformers==4.52.* for this model (naver-hyperclovax/HyperCLOVAX-SEED-Vision-Instruct-3B).
```
WIP :)

dhkim0225 replied:
Updated!
https://huggingface.co/naver-hyperclovax/HyperCLOVAX-SEED-Vision-Instruct-3B/commit/863f35194b15c8dcea9fa0685db1765efb91b567

bigshanedogg committed:
feat: add hyperclovax_seed_vision_3b

Signed-off-by: bigshanedogg <bigshane319@gmail.com>
